include/boost/math/special_functions/prime_sieve.hpp

```diff
@@ -0,0 +1,303 @@
+// Copyright 2020 Matt Borland
+//
+// Use, modification and distribution are subject to the
+// Boost Software License, Version 1.0.
+// (See accompanying file LICENSE_1_0.txt
+// or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef BOOST_MATH_SPECIAL_FUNCTIONS_PRIME_SIEVE_HPP
+#define BOOST_MATH_SPECIAL_FUNCTIONS_PRIME_SIEVE_HPP
+
+#include <boost/math/special_functions/prime.hpp>
+#include <boost/math/special_functions/interval_sieve.hpp>
+#include <boost/assert.hpp>
+#include <vector>
+#include <iterator>
+#include <cmath>
+#include <thread>
+#include <memory>
+#include <future>
+#include <numeric>
+#include <iostream>
+#include <algorithm>
+#include <execution>
+
+namespace boost::math { namespace detail
+{
+// https://mathworld.wolfram.com/SieveofEratosthenes.html
+// https://www.cs.utexas.edu/users/misra/scannedPdf.dir/linearSieve.pdf
+template<class Integer, class Container>
+void linear_sieve(Integer upper_bound, Container &resultant_primes)
+{
+    std::size_t least_divisors_size{static_cast<std::size_t>(upper_bound + 1)};
+    std::unique_ptr<Integer[]> least_divisors{new Integer[least_divisors_size]{0}};
+
+    for (std::size_t i{2}; i < upper_bound; ++i)
+    {
+        if (least_divisors[i] == 0)
+        {
+            least_divisors[i] = i;
+            resultant_primes.emplace_back(i);
+        }
```

Comment:
I would wager that this is a factor in the performance difference: Kim's implementation estimates the number of primes and reserves that number of elements upfront.
I didn't look into exactly how it is estimated but presumably it is something like this: https://en.wikipedia.org/wiki/Prime-counting_function

Reply:
Yes; I would agree-I thought we had this earlier . . .

Reply:
I did. The snippets in question are below:

```
template<typename T>
struct IsVector
{
    using type = T;
    constexpr static bool value = false;
};

template<typename T>
struct IsVector<std::vector<T>>
{
    using type = std::vector<T>;
    constexpr static bool value = true;
};

template<typename T>
constexpr bool is_vector_v = IsVector<T>::value;
```
and this was in the main `prime_sieve` (now callable from `prime_reserve`)

```
    if constexpr (detail::is_vector_v<decltype(primes)>)
    {
        primes.reserve(upper_bound / std::log(static_cast<double>(upper_bound)));
    }
```
I thought making it a little less generic to gain performance was worthwhile.

Reply:
Call me a purist, but you'll make it faster by making it _more_ generic. Allocating memory for the output is not the responsibility of the algorithm.
Now that I've seen KW's implementation, I guess you essentially copied the interface so that it was easier to compare performance?
We don't have to copy the interface, ours can be _better_. And performance is not even the most important factor: allocating memory on the heap is simply a no-go in some contexts.
To make it concrete, if the system that I work on at work needed to generate primes, we could not use KW's or this implementation (as it is), we would have to write our own. I wasn't always aware of such systems and limitations, it just comes with experience.

Reply:
I initially was returning via output iterator instead of using a provided container. Profiling showed that more than 10% of my runtime was the call to `std::move` for the output iterator so I changed approaches to the current one.

Reply:
Wow, `std::move` should be basically free.

What were the compiler options?

Reply:
I believe I used my benchmark options which are "-Ofast -march=native -g -MMD -lgmp -lbenchmark"

Reply:
Could you go back to that commit and get the perf data? Maybe make a [flamegraph](https://github.com/brendangregg/FlameGraph) and post a screenshot of the assembly.

Reply:
Let me edit the current interface to return via output iterator. That way I can be sure there were no confounding variables.

Reply:
Here is the benchmark using the output iterator:

```
-----------------------------------------------------------------------------------------
Benchmark                                               Time             CPU   Iterations
-----------------------------------------------------------------------------------------
prime_sieve<int64_t>/2/real_time                     27.4 ns         27.4 ns     25408214
prime_sieve<int64_t>/4/real_time                     71.5 ns         71.5 ns      9734505
prime_sieve<int64_t>/8/real_time                      111 ns          111 ns      6364109
prime_sieve<int64_t>/16/real_time                     143 ns          143 ns      4894335
prime_sieve<int64_t>/32/real_time                     211 ns          211 ns      3471114
prime_sieve<int64_t>/64/real_time                     287 ns          287 ns      2376086
prime_sieve<int64_t>/128/real_time                    418 ns          418 ns      1675569
prime_sieve<int64_t>/256/real_time                    724 ns          724 ns       960573
prime_sieve<int64_t>/512/real_time                   1239 ns         1238 ns       566861
prime_sieve<int64_t>/1024/real_time                  2442 ns         2440 ns       282865
prime_sieve<int64_t>/2048/real_time                  4990 ns         4987 ns       138413
prime_sieve<int64_t>/4096/real_time                 10807 ns        10799 ns        64025
prime_sieve<int64_t>/8192/real_time                 35965 ns        15096 ns        19436
prime_sieve<int64_t>/16384/real_time                57982 ns        18296 ns        12024
prime_sieve<int64_t>/32768/real_time               114760 ns        21607 ns         6069
prime_sieve<int64_t>/65536/real_time               250246 ns        49289 ns         2809
prime_sieve<int64_t>/131072/real_time              500623 ns        84600 ns         1364
prime_sieve<int64_t>/262144/real_time              995925 ns       154794 ns          668
prime_sieve<int64_t>/524288/real_time             1187309 ns       267116 ns          592
prime_sieve<int64_t>/1048576/real_time            1865367 ns       435442 ns          376
prime_sieve<int64_t>/2097152/real_time            2296991 ns       367161 ns          305
prime_sieve<int64_t>/4194304/real_time            5657749 ns      1538171 ns          120
prime_sieve<int64_t>/8388608/real_time            9159106 ns      2472841 ns           74
prime_sieve<int64_t>/16777216/real_time          16958981 ns      4802018 ns           40
prime_sieve<int64_t>/33554432/real_time          31315935 ns      5518799 ns           23
prime_sieve<int64_t>/67108864/real_time          64255566 ns     12041533 ns           10
prime_sieve<int64_t>/134217728/real_time        133253598 ns     24174532 ns            5
prime_sieve<int64_t>/268435456/real_time        253561668 ns     48189439 ns            3
prime_sieve<int64_t>/536870912/real_time        530443028 ns     95115274 ns            1
prime_sieve<int64_t>/1073741824/real_time       993925796 ns    189247202 ns            1
prime_sieve<int64_t>/real_time_BigO                  0.94 N          0.18 N    
prime_sieve<int64_t>/real_time_RMS                      8 %             4 %   
```
For reference building in the container was :
```
prime_sieve<int64_t>/real_time_BigO                  0.78 N          0.03 N    
prime_sieve<int64_t>/real_time_RMS                     16 %             6 %  
```

Reply:
And here is the relevant section from the call graph. The calls to `std::move` and `std::back_inserter` are not trivial.
![move](https://user-images.githubusercontent.com/3745830/91650129-59e25380-ea41-11ea-9dce-617a52e86f7c.png)

Reply:
Did you compile with `-fno-omit-frame-pointer`? I think these all might get inlined if you don't add that flag. (Obviously the `perf` data will get screwed up, but the google benchmark will be ok.)

Reply:
I did not. `perf record --call-graph dwarf` works without frame pointers.

Reply:
Ok, I think this is right on the boundary of a performance penalty that's acceptable if the ergonomics are improved.

Personally I'm willing to approve merge it either way; but let's hear what other people have to say.

Reply:
I am not convinced this interface change is worth ~5x more CPU runtime. Care to weigh in @pabristow @jzmaddock ?

Reply:
I think we have a number of issues here:

* Using a container mandates dynamic memory allocation which may be a no-go in some contexts, and excludes things like std::array.
* Using an output iterator is fundamentally unsafe in some contexts (ie where the output iterator is a pointer to a C array).
* Using an output iterator on a container is non-trivial as it has to check and resize the container on each output :(

Question: how far can we go with a range-based output?  As in "put the next N primes here".  The range could either be a pair of iterators, an iterator and the value N, or a boost::range.  If that's not faster than a container based interface (no memory allocation) then there's something wrong somewhere.

But... I don't know how you would divide that up for multithreading?

Reply:
@mborland 
I can't see the code that is testing the output iterator, but yeah, a 5x difference is not acceptable. But it's also bizarre. Why is `move` being used at all? `int64_t` is a built-in data type, right, I thought it would just be copied? Is there the same huge difference for `int32_t`?

_I_ would test it something like this. Create the vector to the size required outside the loop.
If you're doing something different, I'd be curious to see it and know why.
```
template <typename Integer>
Integer conservative_prime_count(Integer x)
{
    auto const c = 30 * log(113) / 113; // Magic numbers from Wikipedia.
    return floor(c * x / log(x));
}

template <class Integer>
void prime_sieve(benchmark::State& state)
{
    Integer upper = static_cast<Integer>(state.range(0));
    std::vector<Integer> primes(conservative_prime_count(upper));

    for(auto _ : state)
    {
        benchmark::DoNotOptimize(prime_sieve_helper(std::execution::par, upper, begin(primes)));
    }
    state.SetComplexityN(state.range(0));
}
```

Reply:
@jzmaddock What about building a wrapper class that would support range-based output and other functions like last/next value, random value, etc? I think that is currently the best way to go about supporting a wider variety of data structures. As for removing dynamic memory allocation? I think it is in the realm of the possible, but would require massive retooling. You would become highly dependent on mutex locking which is not a trivial operation.

Reply:
@jzmaddock 
- Right. Doesn't matter how fast an algorithm is if it's not actually useful.
- Not sure why this scenario is different (more or less safe) to any other scenario with an output iterator. Clients are responsible for creating and using output iterators safely. They can use a `back_insert_iterator` on `vector`.
- Yes, but that's an issue for the client, not for the algorithm, right?

(Btw, I'm going to use _x_ for prime numbers and `n` for data sizes.)

It's funny, but the output iterator interface is almost a range interface. As you say, either an (iterator, iterator) or an (iterator, `n`) pair define a range. The interface to this algorithm is essentially (iterator, _x_), which means "generate prime numbers up to _x_", which translates to an output range of (iterator, prime_count(_x_)). 
So yeah, an alternative interface is (iterator, `n`), i.e., generate the first `n` prime numbers. Then it's up to the client to calculate `n` from _x_ if they know what _x_ they want to calculate up to.